### PR TITLE
Writers.setHeader semantic changed to replace + Writers.putHeader

### PR DESCRIPTION
**Pull request by xkrt**
Related to https://github.com/SuaveIO/suave/issues/293

### CODE REVIEW

**Comment by xkrt** [comment]
Looks like this is problem with mono implementation of `HttpClient` (which use Suave.Testing).
Proof:

Add to `Example.Program`:
```fsharp
GET >>= path "/mytest" >>= Writers.putHeader "X-Custom-Header" "first" >>= Writers.putHeader "X-Custom-Header" "second" >>= OK "mytest"
```

Testing client code:
```fsharp
var client = new HttpClient();
var request = new HttpRequestMessage(HttpMethod.Get, new Uri("http://test-host:8082/mytest"));
var resp = client.SendAsync(request).Result;
foreach (var header in resp.Headers)
    Console.WriteLine("{0}: {1}", header.Key, string.Join(", ", header.Value));
```

On Windows with target framework 4.0 I got:
```
X-Custom-Header: first, second
Date: Fri, 09 Oct 2015 13:20:13 GMT
Server: Suave/0.0.0.0, (http://suave.io)
```

But under Ubuntu with official `mono-complete` package  4.0.4.1-0xamarin1 I got:
```
Server: Suave/0.0.0.0, (http://suave.io)
Date: Fri, 09 Oct 2015 13:07:45 GMT
X-Custom-Header: second
```

Guys, what to do with this?

**Comment by ademar** [comment]
I'm not sure what to do with that :confused:

**Comment by xkrt** [comment]
I reported issue to mono https://bugzilla.xamarin.com/show_bug.cgi?id=34715

**Comment by haf** [comment]
@xkrt Can you please read it from the socket like we do in some other tests, instead?

**Comment by xkrt** [comment]
@haf done:)

File: src/Suave/Http.fs [hunk 1]
```diff
@@ -109,7 +109,11 @@ module Http =
     open System
 
     let setHeader key value (ctx : HttpContext) =
-      { ctx with response = { ctx.response with headers = (key, value) :: ctx.response.headers } }
+      { ctx with response = { ctx.response with headers = (key, value) :: (ctx.response.headers |> List.filter (fun (k,_) -> k <> key))  } }
+      |> succeed
+
+    let putHeader key value (ctx : HttpContext) =
```
Comment (haf) [inline src/Suave/Http.fs]: How about `addHeader` instead? Put denotes overwriting to me, similar to how `set` works.

File: src/Suave/Http.fsi [hunk 2]
```diff
@@ -91,9 +91,12 @@ module Http =
   /// Functions have signature f :: params... -> HttpContext -> HttpContext.
   module Writers =
 
-    /// Sets a header with the key and value specified
+    /// Adds or replace a headers with the key and value specified
     val setHeader : key:string -> value:string -> WebPart
```
Comment (haf) [inline src/Suave/Http.fsi]: I'm not onboard with the doc change, as it seems more accurate before. How about "Ensures that the `key` header has the given `value`", instead? And for `addHeader` (renamed from `putHeader`): "Adds the header `key` with the given `value` to the list of returned headers, even if that header already exists"

**Comment by haf** [comment]
@xkrt great fix!

File: src/Suave/Http.fsi (hunk 2 above)
Comment (xkrt) [inline src/Suave/Http.fsi]: @haf I agree with both remarks

**Comment by haf** [comment]
@xkrt Awesome, let's merge! =)

**Pull request merged**